### PR TITLE
[STORAGE] Use subdomain-based paths for our IPFS gateway

### PR DESCRIPTION
**Pull request by nessup**
This PR eliminates the use of `ipfs-{n}.thirdwebcdn.com` gateways. See https://github.com/thirdweb-dev/universal-gateway/pull/20 for a more detailed description of this change.

### CODE REVIEW

**Review by jnsdls — COMMENTED** [review]
~this feels like its no longer needed since we just do the 301 rewrite?~

File: .changeset/quiet-ties-clap.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"@thirdweb-dev/storage": patch
+---
+
+Remove ipfs-2.thirdwebcdn.com due to poor domain authority
```
Comment (jnsdls) [inline .changeset/quiet-ties-clap.md:5]: let's remove this changeset (no reason to have 2 changesets for the same "change")

File: packages/storage/src/common/cid.ts [hunk 2]
```diff
@@ -1,8 +1,8 @@
-import { FileOrBufferOrString } from "../types";
-import { DEFAULT_GATEWAY_URLS } from "./urls";
-import { isBufferOrStringWithName } from "./utils";
+import {FileOrBufferOrString} from "../types";
```
Comment (jnsdls) [inline packages/storage/src/common/cid.ts:1]: prettier will likely complain about this, why the formating change?

File: packages/storage/src/core/downloaders/storage-downloader.ts [hunk 3]
```diff
@@ -33,18 +34,44 @@ export class StorageDownloader implements IStorageDownloader {
...
+    const resOrErr =
+      await fetch(resolvedUri)
+        .catch(err => err)
+
+    if (resOrErr.ok) {
+      return resOrErr
+    }
+
+    if (resOrErr instanceof Response) {
+      if (resOrErr.status === 403) {
```
Comment (jnsdls) [inline packages/storage/src/core/downloaders/storage-downloader.ts:52]: gateway now also returns 410 and in that case we should also *not* retry

**Review by jnsdls — COMMENTED** [review]
some notes

File: packages/storage/src/common/utils.ts [hunk 4]
```diff
@@ -93,9 +108,20 @@ export function replaceGatewayUrlWithScheme(
...
-      if (uri.startsWith(url)) {
-        return uri.replace(url, scheme);
+    for (const gatewayUrl of gatewayUrls[scheme]) {
+      // If the url is a tokenized url, we need to convert it to a canonical url
+      // Otherwise, we just need to check if the url is a prefix of the uri
+      if (gatewayUrl.includes("{cid}")) {
+        // Given the url is a tokenized url, we need to lift the cid and the path from the uri
+        const parsed = parseCidAndPath(gatewayUrl, uri);
+        if (parsed?.cid && parsed?.path) {
+          return `${scheme}${parsed?.cid}/${parsed?.path}`
```
Comment (jakeloo) [inline packages/storage/src/common/utils.ts:118]: Query params being discarded

File: packages/storage/src/core/downloaders/storage-downloader.ts (hunk 3 above)
Comment (jakeloo) [inline packages/storage/src/core/downloaders/storage-downloader.ts:52]: 410

Comment (nessup) [reply packages/storage/src/core/downloaders/storage-downloader.ts:52]: Oof, great catch, thanks1

File: packages/storage/src/common/cid.ts (hunk 2 above)
Comment (nessup) [reply packages/storage/src/common/cid.ts]: Nice catch, fixed!

File: .changeset/quiet-ties-clap.md (hunk 1 above)
Comment (nessup) [reply .changeset/quiet-ties-clap.md:5]: Done

Comment (jnsdls) [reply .changeset/quiet-ties-clap.md:5]: you... kept this one? wasn't the other one better?

**Pull request merged**